docs/api_docs.rst

```diff
@@ -0,0 +1,316 @@
+
+This API consists of a number of object-oriented functor interfaces for saliency heatmap generation.
+These initial interfaces focus on black-box visual saliency.
+We define the two high-level requirements for this initial task: reference image perturbation in preparation for
+black-box testing, and saliency heatmap generation utilizing black-box inputs.
+We define a few similar interfaces for performing the saliency heatmap generation, separated by the intermediate
+algorithmic use-cases: image similarity, classification and object detection.
+We explicitly do not require an abstraction for the black-box operations to fit inside.
+This is intended to allow for applications using these interfaces while leveraging existing functionality, only needing
+to perform any data formatting to fit the input defined here.
+Note, however, that some interfaces are defined for certain black-box concepts as part of the SMQTK ecosystem (e.g.
+in `SMQTK-Descriptors <https://github.com/Kitware/SMQTK-Descriptors>`_, `SMQTK-Classifier <https://github
+.com/Kitware/SMQTK-Classifier>`_, `SMQTK-Relevancy <https://github.com/Kitware/SMQTK-Relevancy>`_ and other SMQTK-*
+modules).
+
+
+These interfaces are based on the plugin and configuration features provided by SMQTK-Core to allow convenient hooks
+into implementation,discoverability, and factory generation from runtime configuration.
+This allows for opaque discovery of interface implementations from a class-method on the interface class object, and
+instantiation of a concrete instance via a JSON-like configuration fed in from an outside resource.
+
+.. figure:: figures/api-docs-fig-01.png
+
+   Figure 1: Abstract Interface Inheritance.
+
+---------------------------
+Perturbed Image Generation
+---------------------------
+
+The PerturbImage interface abstracts the behavior of taking a reference image and generating some number perturbations
+of the image along with paired mask matrices indicating where perturbations have occurred and to what amount.
+
+Implementations should impart no side effects upon the input image.
+
+Immediate candidates for implementation of this interface are occlusion-based saliency algorithms [3] that perform
+perturbations on image pixels.
+
+Interface: PerturbImage
+------------------------
+.. autoclass:: xaitk_saliency.interfaces.perturb_image.PerturbImage
+   :members:
+
+-----------------------------------------
+Image Occlusion via Perturbation Masks
+-----------------------------------------
+
+A common intermediate step in this process is applying the generated perturbation masks to imagery to produce occluded
+images.
+We provide two utility functions as baseline implementation to perform this step:
+
+* ``xaitk_saliency.utils.masking.occlude_image_batch`` - performs the transformation as a batch operation
+
+* ``xaitk_saliency.utils.masking.occlude_image_streaming`` - performs the transformation in a streaming method with
+optional parallelization in that streaming
```

Comment:
This appears as an extra new line in the actual generated docs: see [here](https://xaitk-saliency--51.org.readthedocs.build/en/51/api_docs.html#image-occlusion-via-perturbation-masks)